File: public/scripts/move_localizer.js
export default class MoveLocalizer {


    constructor(locale) {
      this.locale = locale;
      this.CASTLING_REGEX = /^(O-O-O|O-O)([+#])?$/;
      this.STANDARD_MOVE_REGEX = /^([KQRBN])?([a-h])?([1-8])?(x)?([a-h][1-8])(=[QRBN])?([+#])?$/;
      this.CAPTURE = 'x'
      this.PROMOTION = '='
      this.CHECK = '+'
      this.CHECKMATE = '#'

      this.TRANSLATIONS = {
        en: {
          'piece.K': 'K',
          'piece.Q': 'Q',
          'piece.R': 'R',
          'piece.B': 'B',
          'piece.N': 'N',
          'piece.P': '',
        },
        de: {
          'piece.K': 'K',
          'piece.Q': 'D',
          'piece.R': 'T',
          'piece.B': 'L',
          'piece.N': 'S',
          'piece.P': '',
        },
        es: {
          'piece.K': 'R',
          'piece.Q': 'D',
          'piece.R': 'T',
          'piece.B': 'A',
          'piece.N': 'C',
          'piece.P': '',
        },
        fr: {
          'piece.K': 'R',
          'piece.Q': 'D',
          'piece.R': 'T',
          'piece.B': 'L',
          'piece.N': 'S',
          'piece.P': '',
        },
        nb: {
          'piece.K': 'K',
          'piece.Q': 'D',
          'piece.R': 'T',
          'piece.B': 'L',
          'piece.N': 'S',
          'piece.P': '',
        },
        ru: {
          'piece.K': 'Кр',
          'piece.Q': 'Ф',
          'piece.R': 'Л',
          'piece.B': 'С',
          'piece.N': 'К',
          'piece.P': '',
          'file.a': 'а',
          'file.b': 'б',
          'file.c': 'в',
          'file.d': 'г',
          'file.e': 'д',
          'file.f': 'е',
          'file.g': 'ж',
          'file.h': 'з',
        }
      };
    }

    localize(algebraic) {
      // Convert English algebraic notation to moveInfo
      const moveInfo = this.moveInfoFromAlgebraic(algebraic);

      // Convert moveInfo to localized notation
      const piece = moveInfo['piece'] || 'P';
      const pieceLetter = (piece && piece.toUpperCase() != 'P') ? this.TRANSLATIONS[this.locale][`piece.${piece.toUpperCase()}`] : '';

      if (moveInfo['castling']) {
        return algebraic;
      } else {
        // For regular moves, maintain algebraic notation format
        const disambiguation = this.localizeDisambiguation(moveInfo['disambiguation'] || '');  // Use the original disambiguation if any
        const capture = moveInfo['capture'] ? this.CAPTURE : '';
        const to = this.localizeTo(moveInfo['to']); 
        const promotion = moveInfo['promotion'] ? `${this.PROMOTION}${this.TRANSLATIONS[this.locale][`piece.${moveInfo['promotion'].toUpperCase()}`]}` : '';
        const check = moveInfo['check'] ? this.CHECK : '';
        const checkmate = moveInfo['checkmate'] ? this.CHECKMATE : '';

        return `${pieceLetter}${disambiguation}${capture}${to}${promotion}${check}${checkmate}`;
      }
    }

    localizeDisambiguation(disambiguation) {
        if (this.locale !== 'ru') {
            return disambiguation;
        }

        if (disambiguation && disambiguation.match(/^([a-h])([1-8])?$/)) {
            return this.TRANSLATIONS[this.locale][`file.${disambiguation[0]}`] + (disambiguation[1] || '');
        } else {
            return disambiguation;
        }
    }

    localizeTo(to) {
      if (to.length > 1 && this.locale === 'ru') {
        return `${this.TRANSLATIONS[this.locale][`file.${to[0]}`]}${to[1]}`;
      } else {
        return to;
      }
    }

    moveInfoFromAlgebraic(algebraic) {
      // Handle castling first
      const castlingMatch = algebraic.match(this.CASTLING_REGEX);
      if (castlingMatch) {
        const castling = castlingMatch[1] === 'O-O' ? 'kingside' : 'queenside';
        const check = castlingMatch[2] === '+';
        const checkmate = castlingMatch[2] === '#';
        return {
          castling: castling,
          check: check,
          checkmate: checkmate
        };
      }

      // Parse the move components
      const standardMatch = algebraic.match(this.STANDARD_MOVE_REGEX);
      if (standardMatch) {
        const piece = standardMatch[1] || 'P';
        const fileHint = standardMatch[2];
        const rankHint = standardMatch[3];
        const capture = standardMatch[4] === 'x';
        const to = standardMatch[5];
        const promotion = standardMatch[6]?.replace('=', '');
        const check = standardMatch[7] === '+';
        const checkmate = standardMatch[7] === '#';

        // Keep original disambiguation (if any)
        let disambiguation = '';
        if (fileHint) { disambiguation += fileHint; }
        if (rankHint) { disambiguation += rankHint; }

        return {
          piece: piece,
          disambiguation: disambiguation,
          to: to,
          capture: capture,
          promotion: promotion,
          check: check,
          checkmate: checkmate
        };
      } else {
        throw new Error(`Invalid algebraic notation: ${algebraic}`);
      }
    }

}

File: lib/move_localizer.rb
# frozen_string_literal: true

require 'i18n'

class MoveLocalizer
  CAPTURE = 'x'
  PROMOTION = '='
  CHECK = '+'
  CHECKMATE = '#'

  STANDARD_MOVE_REGEX = /^([KQRBN])?([a-h])?([1-8])?(x)?([a-h][1-8])(=[QRBN])?([+#])?$/
  CASTLING_REGEX = /^O-O(-O)?([+#])?$/

  def initialize(locale)
    @locale = locale
  end

  def localize_move(algebraic)
    move_info = move_info_from_algebraic(algebraic)
    text = move_info[:castling] ? algebraic : build_localized_move(move_info)
    { text: text }
  end

  private

  def localize_disambiguation(disambiguation)
    return disambiguation if @locale.to_s != 'ru'

    if disambiguation&.match(/^([a-h])([1-8])?$/)
      return I18n.t("chess.files.#{disambiguation[0]}", locale: @locale) + (disambiguation[1] || '')
    end

    disambiguation
  end

  def localize_to(to)
    return "#{I18n.t("chess.files.#{to[0]}", locale: @locale)}#{to[1]}" if to.length > 1 && @locale.to_s == 'ru'

    to
  end

  def move_info_from_algebraic(algebraic)
    return parse_castling_move(algebraic) if CASTLING_REGEX.match?(algebraic)

    parse_standard_move(algebraic)
  end

  def parse_castling_move(algebraic)
    match = CASTLING_REGEX.match(algebraic)
    {
      castling: match[1] ? 'queenside' : 'kingside',
      check: match[2] == '+',
      checkmate: match[2] == '#'
    }
  end

  def parse_standard_move(algebraic)
    match = STANDARD_MOVE_REGEX.match(algebraic)
    raise "Invalid algebraic notation: #{algebraic}" unless match

    move_components = extract_move_components(match)
    move_components[:file_hint] = build_disambiguation(move_components)
    move_components
  end

  def extract_move_components(match)
    {
      piece: match[1] || 'P',
      to: match[5],
      capture: match[4] == 'x',
      promotion: extract_promotion(match[6]),
      check: match[7] == '+',
      checkmate: match[7] == '#',
      file_hint: match[2],
      rank_hint: match[3]
    }
  end

  def extract_promotion(promotion_string)
    promotion_string&.sub('=', '')
  end

  def build_disambiguation(components)
    disambiguation = ''
    disambiguation += components[:file_hint] if components[:file_hint]
    disambiguation += components[:rank_hint] if components[:rank_hint]
    disambiguation
  end

  def build_localized_move(move_info)
    [
      get_piece_letter(move_info[:piece]),
      localize_disambiguation(move_info[:file_hint]),
      get_capture_symbol(move_info[:capture]),
      localize_to(move_info[:to]),
      build_promotion(move_info[:promotion]),
      build_check_symbols(move_info)
    ].join
  end

  def get_piece_letter(piece)
    return '' unless piece && piece.upcase != 'P'

    I18n.t("chess.pieces.#{piece.upcase}", locale: @locale)
  end

  def get_capture_symbol(capture)
    capture ? CAPTURE : ''
  end

  def build_promotion(promotion)
    return '' unless promotion

    "#{PROMOTION}#{I18n.t("chess.pieces.#{promotion.upcase}", locale: @locale)}"
  end

  def build_check_symbols(move_info)
    return CHECKMATE if move_info[:checkmate]
    return CHECK if move_info[:check]

    ''
  end
end
